internal/langserver/handlers/initialize.go

```diff
@@ -134,6 +134,18 @@ func (svc *service) Initialize(ctx context.Context, params lsp.InitializeParams)
 		rootURI := string(params.RootURI)
 		if !uri.IsURIValid(rootURI) {
 			properties["root_uri"] = "invalid"
+
+			if uri.IsURLEncodedPath(rootURI) {
+				uri, err := uri.UnEncodeURI(rootURI)
+				if err == nil {
+					if uri.Scheme == "file" && uri.Host == "wsl$" {
+						// panic("foo")
+						return serverCaps, fmt.Errorf("Unsupported WSLPATH: %q "+
+							"This is most likely bug, please report it.", rootURI)
+					}
+				}
+			}
+
 			return serverCaps, fmt.Errorf("Unsupported or invalid URI: %q "+
 				"This is most likely bug, please report it.", rootURI)
```

Comment:
```suggestion
			invalidUriErr := jrpc2.Errorf(code.InvalidParams,
				"Unsupported or invalid URI: %q "+
					"This is most likely client bug, please report it.", rootURI)

			if uri.IsWSLURI(rootURI) {
				// For WSL URIs we return additional error data
				// such that clients (e.g. VS Code) can provide better UX
				// and nudge users to open the Linux path instead.
				return serverCaps, invalidUriErr.WithData("INVALID_URI_WSL")
			}

			return serverCaps, invalidUriErr
```

The `Data` field can be anything per LSP - I assume that string is easy to deal with on client-side and also easy to build here, but feel free to change it to anything else if you feel it's more convenient.

https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#responseError

I also proposed updating of the error for the generic case when the URI is not WSL but it's still invalid, as I think `InvalidParams` is more appropriate/relevant than the default `ServerError`.